docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,552 @@
+rules such that the rule with higher priority (match could be generic) will override the rule
+with lower priority (match will be specific. See: Story-3).
+
+The CRD will be Namespaced, with multiple resources
+allowed per namespace. The resources will be watched by OVN-K8s, which in turn will configure
+OVN's [QoS Table](https://man7.org/linux/man-pages/man5/ovn-nb.5.html#NetworkQoS_TABLE). The  
+`NetworkQoS` also has `status` field which is populated by OVN-K8s which helps users to identify
+whether NetworkQoS rules are configured correctly in OVN or not.
+
+### API Details
+
+* A new API `NetworkQoS` under the `k8s.ovn.org/v1alpha1` group will be added to  
+  `go-controller/pkg/crd/networkqos/v1alpha1`. This would be a namespace-scoped CRD:
+
+```go
+import (
+corev1 "k8s.io/api/core/v1"
+networkingv1 "k8s.io/api/networking/v1"
+metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=networkqoses
+// +kubebuilder::singular=networkqos
+// +kubebuilder:object:root=true
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=".status.status"
+// +kubebuilder:subresource:status
+// NetworkQoS is a CRD that allows the user to define a DSCP marking and metering
+// for pods ingress/egress traffic on its namespace to specified CIDRs,
+// protocol and port. Traffic belong these pods will be checked against
+// each Rule in the namespace's NetworkQoS, and if there is a match the traffic
+// is marked with relevant DSCP value and enforcing specified policing/shaping
+// parameters.
+type NetworkQoS struct {
+metav1.TypeMeta   `json:",inline"`
+metav1.ObjectMeta `json:"metadata,omitempty"`
+
+Spec   Spec   `json:"spec,omitempty"`
+Status Status `json:"status,omitempty"`
+}
+
+// Spec defines the desired state of NetworkQoS
+type Spec struct {
+// netAttachRefs points to a list of objects which could be either NAD, UDN, or Cluster UDN.
+// In the case of NAD, the network type could be of type Layer-3, Layer-2, or Localnet.
+// If not specified, then the primary network of the selected Pods will be chosen.
+// +optional
+// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="netAttachRefs is immutable"
+NetworkAttachmentRefs []corev1.ObjectReference `json:"netAttachRefs,omitempty"`
+
+// podSelector applies the NetworkQoS rule only to the pods in the namespace whose label
+// matches this definition. This field is optional, and in case it is not set
+// results in the rule being applied to all pods in the namespace.
+// +optional
+PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+
+// priority is a value from 0 to 100 and represents the NetworkQoS' priority.
+// QoSes with numerically higher priority takes precedence over those with lower.
+// +kubebuilder:validation:Maximum:=1
+// +kubebuilder:validation:Minimum:=100
+Priority int `json:"priority"`
```

Comment:
Do we want to follow ANP design here, and make priority with lower value take precedence?
So that priority 1 is applied before priority 2? This is optional, but wanted to point this out